### PR TITLE
Pat.Meta in Pat.Meta

### PR DESCRIPTION
**Pull request by HoshinoTented**
This PR allows this code compiles:

```aya
open data Nat | O | S Nat
open data INat (n : Nat)
| O => zero
| S n' => +-one
| S (S n') => +-two

def yes {n : Nat} (a : INat n) (b : INat n) : Nat
| +-one, +-two => O
| _, _ => S O
```

The `yes` function is equivalent to:

```aya
def yes {n : Nat} (a : INat n) (b : INat n) : Nat
| {S (S n')}, +-one, +-two => O
| {n}, _, _ => S O
```

### CODE REVIEW

File: base/src/main/java/org/aya/core/visitor/PatTraversal.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+  }
+
+  default @NotNull Pat descent(@NotNull Pat pat) {
+    return switch (pat) {
+      case Pat.Absurd absurd -> absurd;
+      case Pat.Bind bind -> bind;
+      case Pat.Ctor ctor -> {
+        var params = ctor.params().map(this);
+
+        if (params.sameElements(ctor.params())) yield ctor;
```
Comment (ice1000) [inline base/src/main/java/org/aya/core/visitor/PatTraversal.java:35]: `, true`

File: base/src/main/java/org/aya/core/visitor/PatTraversal.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+          yield new Pat.Meta(meta.explicit(), MutableValue.create(solution), meta.fakeBind(), meta.type());
+        } else {
+          yield meta;
+        }
+      }
+      case Pat.ShapedInt shapedInt -> shapedInt;
+      case Pat.Tuple tuple -> {
+        var pats = tuple.pats().map(this);
+
+        if (pats.sameElements(tuple.pats())) yield tuple;
```
Comment (ice1000) [inline base/src/main/java/org/aya/core/visitor/PatTraversal.java:54]: Ditto

File: base/src/main/java/org/aya/core/visitor/PatTraversal.java (hunk 1 above)
Comment (HoshinoTented) [reply base/src/main/java/org/aya/core/visitor/PatTraversal.java]: Kala's api changed?

File: base/src/main/java/org/aya/core/pat/Pat.java [hunk 3]
```diff
@@ -48,6 +48,8 @@ public sealed interface Pat extends AyaDocile {
   @Override default @NotNull Doc toDoc(@NotNull DistillerOptions options) {
     return new CoreDistiller(options).pat(this, BaseDistiller.Outer.Free);
   }
+
+  // TODO[hoshino]: useless now
   @NotNull Pat rename(@NotNull Subst subst, @NotNull LocalCtx localCtx, boolean explicit);
```
Comment (ice1000) [inline base/src/main/java/org/aya/core/pat/Pat.java:53]: Care to delete it?

**Review by ice1000 — APPROVED** [review]
Rest lgtm, feel free to merge

**Pull request merged**